src/metatrain/utils/data/dataset.py

```diff
@@ -232,17 +236,39 @@ def get_all_targets(datasets: Union[Dataset, List[Dataset]]) -> List[str]:
     return sorted(set(target_names))
 
 
-def collate_fn(batch: List[Dict[str, Any]]) -> Tuple[List, Dict[str, TensorMap]]:
+def collate_fn(
+    batch: List[Dict[str, Any]],
+) -> Tuple[List, Dict[str, TensorMap], Dict[str, TensorMap]]:
     """
-    Wraps `group_and_join` to
-    return the data fields as a list of systems, and a dictionary of nameed
-    targets.
+    Wraps `group_and_join` to return the data fields as a list of systems, and a
+    dictionary of named targets.
+
+    The ``system`` field is treated specially, and is returned as the first element in
+    the returned tuple. All other fields whose names do not start with "extra::" are
+    considered to be targets and returned in a dictionary as the second element of the
+    returned tuple. Any fields whose names start with "extra::" are considered to be
+    extra data, and are returned in a separate dictionary as the third element of the
+    returned tuple.
```

Comment:
Instead of making a distinction on whether the name starts with `extra::` or not, could be pass this information separatetly? Maybe using `dataset_info.targets`? This might require turning this collate_fn into a class with a `__call__` function so we can store some state inside.

Reply:
>  But then this requires partial initialization of the collate function (i.e. with functools.partial) which torchscript doesn't like.

Either that or using a class, which might work better than `partial`

> In my mind the collate function should be as simple as possible, and relying on a extra:: prefix achieves this and is easily enforceable.

I disagree: the goal is to be as simple as possible for the **end user**, and the prefix is a bit jarring and redundant here. If we can solve things with a bit more work on the dev side I'm all for it. 

Also why would this function run in TorchScript mode? It should never be part of a model's `forward`.